R/rows-d.R

```diff
@@ -0,0 +1,214 @@
+#' Persisting data
+#'
+#' @description
+#' \lifecycle{experimental}
+#'
+#' These methods provide a framework for manipulating individual rows
+#' in existing tables, modeled after the SQL operations
+#' `INSERT`, `UPDATE` and `DELETE`.
+#' All operations expect existing and new data to be compatible.
+#'
+#' @param .data Target table object.
+#' @param .key Key columns, unquoted.
+#'   The default varies for the different operations:
+#'   - For `rows_insert()`: no keys.
+#'   - For `rows_update()`, `rows_patch()` and `rows_upsert()`: first column.
+#'   - For `rows_delete()`: all columns.
+#' @param ... New data.
+#'   If unnamed, must be a single object of the same class as `.data`.
+#'   If named, will be passed on to [tibble()].
+#'
+#' @return A tbl object of the same structure as `.data`.
+#'
+#' @name rows
+#' @example example/rows-d.R
+NULL
+
+
+#' rows_insert
+#'
+#' `rows_insert()` adds new rows.
+#' This operation corresponds to `INSERT` in SQL.
+#' If `.key` is given, no two rows with the same values in the key columns
+#' are permitted.
+#' @param .sort Should the output be sorted by the key columns?
+#' @rdname rows
+#' @export
+rows_insert <- function(.data, ..., .key = NULL) {
+  UseMethod("rows_insert", .data)
+}
+
+#' @export
+#' @rdname rows
+rows_insert.data.frame <- function(.data, ..., .key = NULL, .sort = NULL) {
+  source <- dots_to_df(.data, ...)
+  source_key <- df_key(source, {{ .key }}, default = !!integer())
+  data_key <- df_key(.data, !!names(source)[source_key])
+
+  out <- vctrs::vec_rbind(.data, source)
+  if (isTRUE(.sort)) {
+    # FIXME: Faster implementation, sort before binding
```

Comment:
I think that's going to be hard to implement because you'd need to sort `x` and `y` individually then merge sort the two together? I doubt that'll be worth the effort.

Reply:
But I think it's odd to mix vctrs and dplyr here — instead of arrange, maybe use `vec_order()`?